lib/GenomeFileUtil/core/GenomeInterface.py

```diff
@@ -128,54 +193,75 @@ def get_one_genome(self, params):
-
-        # check all handles point to shock nodes owned by calling user
-        self._own_handle(data, 'genbank_handle_ref')
-        self._own_handle(data, 'gff_handle_ref')
-        if "AnnotatedMetagenomeAssembly" not in ws_datatype:
-            self._check_dna_sequence_in_features(data)
-            data['warnings'] = self.validate_genome(data)
-
-        # sort data
-        data = GenomeUtils.sort_dict(data)
-        # dump genome to scratch for upload
-        data_path = os.path.join(self.scratch, name + ".json")
-        json.dump(data, open(data_path, 'w'))
-        if 'hidden' in params and str(params['hidden']).lower() in ('yes', 'true', 't', '1'):
-            hidden = 1
-        else:
-            hidden = 0
-
-        if isinstance(workspace, int) or workspace.isdigit():
-            workspace_id = workspace
-        else:
-            workspace_id = self.dfu.ws_name_to_id(workspace)
-
-        save_params = {'id': workspace_id,
-                       'objects': [{'type': ws_datatype,
-                                    'data_json_file': data_path,
-                                    'name': name,
-                                    'meta': meta,
-                                    'hidden': hidden}]}
-        dfu_oi = self.ws_large_data.save_objects(save_params)[0]
-        returnVal = {'info': dfu_oi, 'warnings': data.get('warnings', [])}
-        return returnVal
+    def _save_genome_mass(self, params):
+
+        workspace_id = params[_WSID]
+        inputs = params[_INPUTS]
+
+        ws_datatypes = []
+        data_paths = []
+        names = []
+        meta_data = []
+        hidden_data = []
+        warnings = []
+
+        for input_params in inputs:
+
+            # retrive required params
+            name = input_params['name']
+            data = input_params['data']
+
+            # XXX there is no `workspace_datatype` param in the spec
+            ws_datatype = input_params.get('workspace_datatype', "KBaseGenomes.Genome")
+            # XXX there is no `meta` param in the spec
+            meta = input_params.get('meta', {})
+
+            ws_datatypes.append(ws_datatype)
+            names.append(name)
+            meta_data.append(meta)
+
+            if "AnnotatedMetagenomeAssembly" in ws_datatype:
+                if input_params.get('upgrade') or 'feature_counts' not in data:
+                    data = self._update_metagenome(data)
+            else:
+                if input_params.get('upgrade') or 'feature_counts' not in data:
+                    data = self._update_genome(data)
+
+            # check all handles point to shock nodes owned by calling user
+            self._own_handle(data, 'genbank_handle_ref')
+            self._own_handle(data, 'gff_handle_ref')
+            if "AnnotatedMetagenomeAssembly" not in ws_datatype:
+                self._check_dna_sequence_in_features(data)
+                data['warnings'] = self.validate_genome(data)
```

Comment:
here's another place where the upload could fail after saving data

Reply:
Which line are you referring to here? Line 235?

Reply:
Both of them sound like they could throw errors based on the contents of the genome

Reply:
1. They are called before _save_genome_objects in GenomeInterface.py
2. I don't think they will throw errors (GenomeInterface.handle_large_genomes might raise an error if the genome is too large. However, as we discussed, we are planning to use GTG for batching.)

Reply:
But they're called after assemblies are saved, as so if they do throw an error, we'd have partial saves.

Reply:
_check_dna_sequence_in_features could fail if the assembly ref or contig ref in the genome object is invalid, or if the assembly sequence api call fails

Reply:
validate_genomes appears to throw an error on a single genome so batching earlier won't help

Reply:
👍

Reply:
It looks like these functions are now run on all genomes but the originally only ran on metagenomes?

Reply:
```
if "AnnotatedMetagenomeAssembly" not in ws_datatype:
    self.check_dna_sequence_in_features(data)
    data['warnings'] = self.validate_genome(data)
```

The original logic runs on `KBaseGenomes.Genome`. 
Now it runs `KBaseGenomes.Genome`  as well? 
Correct me if I am wrong, I personally have not seen any `AnnotatedMetagenomeAssembly` got pass into GTG.

Reply:
Oh sorry, missed the `not`. Can you even specify the type for GTG?

Reply:
no, workspace_datatype does not exist in specs

Reply:
But if I specify it in my input will it be used? The spec doesn't prevent additional keys in the dict from being passed into the method.

Reply:
Are there no key checks from the specification? If so, what is the purpose of having specifications?

Reply:
There are key checks in Java, just not Python. The specs provide stub compilation and documentation. That being said, if you want to argue the SDK is crummy in many ways I won't disagree.

Reply:
This is getting tricky because if any keys can be specified in the input parameters, which keys should we check for the type?

Reply:
I feel the line is getting blurry here. `workspace_datatype` was not used properly in GenomeInterface.py and now we need to introduce it in GTG.

Reply:
Whoa, we do *not* want to introduce it into GTG. I'm only asking if it *can* be used in GTG

Reply:
What is the difference between `it can be passed into GTG` and `it can be used in GTG`?

Reply:
Same thing. The question I'm asking if is a user specifies that key in the right place in the GTG input will it be picked up in GenomeInterface and the type set based on the key.

Reply:
Now only `name`, `meta`, `input` got passed into GenomeInterface from GTG
https://github.com/kbaseapps/GenomeFileUtil/blob/71670bce3fd95445172e415c6167a0471cd697f6/lib/GenomeFileUtil/core/GenbankToGenome.py#L199

Reply:
:+1:

Reply:
It all depends on whether a user specifies workspace_datatype in the GTG input parameters. If the user does not specify it, which we assume is the case, the current code is sufficient. However, if the user does specify it for any reason, we will need to add a condition in GTG.

Reply:
Wait, you just said that GTG effectively ignores it. Why would we need to add a condition?

Reply:
zoom?

Reply:
👍

Reply:
Upshot is as above, the data type key is ignored by G2G which is what we want

:+1: